File: midi-script/Namespaces/ChainMixerDevice.py

from .DeviceParameter import DeviceParameter
from .Interface import Interface


class ChainMixerDevice(Interface):
    @staticmethod
    def serialize_chain_mixer_device(mixer_device):
        if mixer_device is None:
            return None

        device_id = Interface.save_obj(mixer_device)

        try:
            volume = mixer_device.volume
        except RuntimeError:
            # MIDI Effect Rack chains have no volume parameter.
            volume = None

        return {"id": device_id, "volume": volume}

    def __init__(self, c_instance, socket):
        super().__init__(c_instance, socket)

    def get_chain_activator(self, ns):
        return DeviceParameter.serialize_device_parameter(ns.chain_activator)

    def get_panning(self, ns):
        try:
            return DeviceParameter.serialize_device_parameter(ns.panning)
        except RuntimeError:
            return None

    def get_sends(self, ns):
        try:
            return list(map(DeviceParameter.serialize_device_parameter, ns.sends))
        except RuntimeError:
            return []

    def get_volume(self, ns):
        try:
            return DeviceParameter.serialize_device_parameter(ns.volume)
        except RuntimeError:
            return None

File: src/ns/chain-mixer-device.ts
import { Ableton } from "../index.js";
import { Namespace } from "./index.js";
import { DeviceParameter, RawDeviceParameter } from "./device-parameter.js";

export interface GettableProperties {
  chain_activator: RawDeviceParameter;
  /** `null` on MIDI Effect Rack chains (no pan). */
  panning: RawDeviceParameter | null;
  sends: RawDeviceParameter[];
  /** `null` on MIDI Effect Rack chains (no volume). */
  volume: RawDeviceParameter | null;
}

export interface TransformedProperties {
  chain_activator: DeviceParameter;
  panning: DeviceParameter | null;
  sends: DeviceParameter[];
  volume: DeviceParameter | null;
}

export interface SettableProperties {}

export interface ObservableProperties {
  sends: RawDeviceParameter[];
}

export interface RawChainMixerDevice {
  readonly id: string;
  /** Display string when present; `null` on MIDI Effect chains. */
  readonly volume: string | null;
}

/**
 * Mixer for a rack chain: chain activator, volume, pan, and sends.
 */
export class ChainMixerDevice extends Namespace<
  GettableProperties,
  TransformedProperties,
  SettableProperties,
  ObservableProperties
> {
  constructor(
    ableton: Ableton,
    public readonly raw: RawChainMixerDevice,
  ) {
    super(ableton, "chain-mixer-device", raw.id);

    this.transformers = {
      chain_activator: (v) => new DeviceParameter(ableton, v),
      panning: (v) => (v ? new DeviceParameter(ableton, v) : null),
      sends: (v) => v.map((s) => new DeviceParameter(ableton, s)),
      volume: (v) => (v ? new DeviceParameter(ableton, v) : null),
    };
  }
}
